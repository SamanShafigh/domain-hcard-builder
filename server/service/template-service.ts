const { createReadStream } = require('fs');
const React = require('react');
const ReactDom = require('react-dom/server')

interface Global {
  React: any;
}

declare var global: Global;
global.React = React;

function html(body: string) {
  return `
  <!DOCTYPE html>
  <html lang="en">
    <head>
      <meta charset="utf-8">
      <meta http-equiv="X-UA-Compatible" content="IE=edge">
      <meta name="viewport" content="width=device-width, initial-scale=1">
      <title>Live hCard Preview</title>
  
      <link href="css/bootstrap.min.css" rel="stylesheet" >
      <link href="css/main.css" rel="stylesheet">
    </head>
  
    <body>
      ${body}
    </body>
  </html>
  `;
} 

function render(path: string, props: any) {
  const component = require(path).default;
  var body = ReactDom.renderToString(
    React.createElement(component, props)
  );

  return html(body);
}

function serve(path: string) {
  return createReadStream(`${__dirname}/${path}`)
}

module.exports = {
  render,
  serve
}