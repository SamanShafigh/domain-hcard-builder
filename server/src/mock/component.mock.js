module.exports = {
  default: 'h1'
}