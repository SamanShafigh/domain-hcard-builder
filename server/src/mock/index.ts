import HCardT from './type';

export const mockEnv = <HCardT.Env>{
  STATIC_ROOT: 'root',
  PORT: 1234,
  RENDER_MODE: 'ssr',
  DB_URL: 'mongodb://localhost',
  DB_NAME: 'test'
};

export const mockConfig = {
  staticRoot: './client/dist',
  ssrPath: './dist/main.js',
  spaPath: './dist/_index.html',
  serverPort: 3030,
  renderMode: 'ssr',
  dbUri: 'mongodb://localhost',
  dbName: 'hcard',
}

export const mockUser = {
  givenName: 'givenName',
  surname: 'surname',
  email: 'email',
  phone: 'phone',
  houseNumber: 'houseNumber',
  street: 'street',
  suburb: 'suburb',
  state: 'state',
  postcode: 'postcode',
  country: 'country'
}

export const mockInvalidUser = {
  givenName: 'givenName',
  email: 'email',
  houseNumber: 'houseNumber',
  suburb: 'suburb',
  postcode: 'postcode',
  invalidData: 'some Invalid data'
}

export const mockCtx = {
  meta: {
    userId: '1234',
    noscript: true,
  }
}

export const mockCtxWithBody = {
  meta: {
    userId: '1234',
    noscript: true,
  },
  request: {
    body: mockUser
  }
}

export const mockDbDriver = {
  findOne: (model, query, options) => (Promise.resolve({ data: mockUser })),
  save: (model, query, data, options) => ({ model, query, data, options }),
  update: (model, query, data, options) => ({ model, query, data, options }),
}

export const mockMongoDbDriver = {
  collection: (model) => ({
    findOne: (query, options) => Promise.resolve({ model, query, options }),
    update: (query, data, options) => Promise.resolve({ model, query, data, options}),
  })
}

export const mockReactComponentPath = '../mock/component.mock.js'
export const mockHtmlPagePath = '../mock/page.mock.html'