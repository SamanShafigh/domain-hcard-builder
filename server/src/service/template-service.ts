import { createReadStream } from 'fs';
import * as React from 'react';
import * as ReactDom from 'react-dom/server';

interface Global { React: any; }
declare var global: Global;
global.React = React;

function html(body: string) {
  return `
  <!DOCTYPE html>
  <html lang="en">
    <head>
      <meta charset="utf-8">
      <meta http-equiv="X-UA-Compatible" content="IE=edge">
      <meta name="viewport" content="width=device-width, initial-scale=1">
      <title>Live hCard Preview</title>
  
      <link href="css/bootstrap.min.css" rel="stylesheet" >
      <link href="css/main.css" rel="stylesheet">
    </head>
  
    <body>
      ${body}
    </body>
  </html>
  `;
} 

export function render(path: string, props: any) {
  const component = require(path).default;
  var body = ReactDom.renderToString(
    React.createElement(component, props)
  );

  return html(body);
}

export function serve(path: string) {
  return createReadStream(`${__dirname}/${path}`)
}
