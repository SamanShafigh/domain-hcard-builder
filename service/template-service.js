const React = require('react');
const ReactDom = require('react-dom/server')
global.React = React;

function html(body) {
  return `
  <!DOCTYPE html>
  <html lang="en">
    <head>
      <meta charset="utf-8">
      <meta http-equiv="X-UA-Compatible" content="IE=edge">
      <meta name="viewport" content="width=device-width, initial-scale=1">
      <title>Live hCard Preview</title>
  
      <link href="css/bootstrap.min.css" rel="stylesheet" >
      <link href="css/main.css" rel="stylesheet">
    </head>
  
    <body>
      ${body}
    </body>
  </html>
  `;
} 

module.exports.render = (component, props) => {
  var body = ReactDom.renderToString(
    React.createElement(component, props)
  );

  return html(body);
}