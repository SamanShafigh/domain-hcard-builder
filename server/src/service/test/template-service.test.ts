import * as TemplateService from '../template-service';
import { mockReactComponentPath, mockHtmlPagePath } from '../../mock';
import * as React from 'react';

describe('Test template service', () => {
  test('test render a react component', async () => {
    const component = TemplateService.render(mockReactComponentPath);
    expect(component).toMatchSnapshot();
  });

  test('test html function', async () => {
    const component = TemplateService.html('some content');
    expect(component).toMatchSnapshot();
  });

  test('test serve a html file', async () => {
    function getContent() {
      return new Promise((resolve, reject) => {
        const chunks = [];
        const stream = TemplateService.serve(mockHtmlPagePath);
        stream.on('data', (chunk) => {
          chunks.push(chunk.toString());
        });
        stream.on('end', () => {
          resolve(chunks.join(''));
        });
        stream.on('error', reject);
      })
    }

    const content = await getContent();
    expect(content).toMatchSnapshot();
  });
});