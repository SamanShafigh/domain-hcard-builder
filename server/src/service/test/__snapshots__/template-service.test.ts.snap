// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test template service test html function 1`] = `
"
  <!DOCTYPE html>
  <html lang=\\"en\\">
    <head>
      <meta charset=\\"utf-8\\">
      <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
      <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
      <title>Live hCard Preview</title>
  
      <link href=\\"css/bootstrap.min.css\\" rel=\\"stylesheet\\" >
      <link href=\\"css/main.css\\" rel=\\"stylesheet\\">
    </head>
  
    <body>
      some content
    </body>
  </html>
  "
`;

exports[`Test template service test render a react component 1`] = `
"
  <!DOCTYPE html>
  <html lang=\\"en\\">
    <head>
      <meta charset=\\"utf-8\\">
      <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
      <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
      <title>Live hCard Preview</title>
  
      <link href=\\"css/bootstrap.min.css\\" rel=\\"stylesheet\\" >
      <link href=\\"css/main.css\\" rel=\\"stylesheet\\">
    </head>
  
    <body>
      <h1 data-reactroot=\\"\\" data-reactid=\\"1\\" data-react-checksum=\\"1245974022\\"></h1>
    </body>
  </html>
  "
`;

exports[`Test template service test serve a html file 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
<head>
  <meta charset=\\"UTF-8\\">
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
  <meta http-equiv=\\"X-UA-Compatible\\" content=\\"ie=edge\\">
  <title>Document</title>
</head>
<body>
  This is for test only
</body>
</html>"
`;
